src/main/java/com/microsoft/aad/msal4j/TokenRequest.java

```diff
@@ -139,40 +130,50 @@ AuthenticationResult executeOauthRequestAndProcessResponse()
-                ErrorObject errorObject = errorResponse.getErrorObject();
 
-                if (AuthenticationErrorCode.AUTHORIZATION_PENDING.toString()
-                        .equals(errorObject.getCode())) {
-
-                    httpEvent.setOauthErrorCode(AuthenticationErrorCode.AUTHORIZATION_PENDING.toString());
-
-                    throw new AuthenticationException(AuthenticationErrorCode.AUTHORIZATION_PENDING,
-                            errorObject.getDescription());
+                String responseContent = httpResponse.getContent();
+                if(responseContent == null || StringHelper.isBlank(responseContent)){
+                    throw new AuthenticationServiceException("Unknown Service Exception");
                 }
 
-                if (HTTPResponse.SC_BAD_REQUEST == errorObject.getHTTPStatusCode() &&
-                        AuthenticationErrorCode.INTERACTION_REQUIRED.toString().equals(errorObject.getCode())) {
+                ErrorResponse errorResponse = JsonHelper.convertJsonToObject(
+                        responseContent,
+                        ErrorResponse.class);
 
-                    httpEvent.setOauthErrorCode(AuthenticationErrorCode.INTERACTION_REQUIRED.toString());
+                errorResponse.statusCode(httpResponse.getStatusCode());
+                errorResponse.statusMessage(httpResponse.getStatusMessage());
 
-                    throw new ClaimsChallengeException(
-                            errorResponse.toJSONObject().toJSONString(),
-                            getClaims(httpResponse.getContent()));
-                } else {
-                    String telemetryErrorCode = StringHelper.isBlank(errorObject.getCode()) ?
-                            AuthenticationErrorCode.UNKNOWN.toString() :
-                            errorObject.getCode();
+                // Some invalid_grant or interaction_required subError codes returned by
+                // the service are not supposed to be exposed to customers
+                if(errorResponse.error() != null &&
+                        errorResponse.error().equalsIgnoreCase(AuthenticationErrorCode.INVALID_GRANT) ||
+                        errorResponse.error().equalsIgnoreCase(AuthenticationErrorCode.INTERACTION_REQUIRED)){
+                    errorResponse = filterSubErrorCode(errorResponse);
+                }
 
-                    httpEvent.setOauthErrorCode(telemetryErrorCode);
+                httpEvent.setOauthErrorCode(errorResponse.error());
 
-                    throw new AuthenticationException(errorResponse.toJSONObject().toJSONString());
-                }
+                throw new AuthenticationServiceException(
+                        errorResponse,
+                        httpResponse.getHeaderMap());
             }
             return result;
         }
     }
 
+    private ErrorResponse filterSubErrorCode(ErrorResponse errorResponse){
```

Comment:
The spec I've used to do this work mentioned having a "Classification" property that exposes some of the sub-errors. We don't expose sub_error in general. 

Also, for discoverability, the Classification property will return an enum instead of a string constant.